src/Chapter4.hs

```diff
@@ -489,6 +500,20 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
+instance Applicative List where
+    pure :: a -> List a
+    pure a = Cons a (Empty)
+
+    (<*>) :: List (a -> b) -> List a -> List b
+    Empty <*> _ = Empty
+    _ <*> Empty = Empty
+    Cons f a <*> c = joinLists (fmap f c) (a <*> c)
+
+joinLists :: List a -> List a -> List a
+joinLists Empty c = c
+joinLists (Cons a b) c = Cons a (joinLists b c)
```

Comment:
This is the correct implementation! You can implement it even more efficiently by pattern-matching on the second list and return the first one immediately if the second is `Empty`. So you will avoid traversing the first list if the second is empty.

Usually, such function is called `append`. In Haskell, the standard typeclass for appending things is called `Semigroup`. So, if you want to practice more, you can implement instance of `Semigroup` for `List` and use it in the `Applicative` instance 🙂 

* http://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Semigroup.html#t:Semigroup